cmake/add_enclave.cmake

```diff
@@ -101,3 +147,91 @@ function(add_enclave)
 endfunction()
+
+macro(add_enclave_optee)
+   set(options CXX)
+   set(oneValueArgs TARGET UUID KEY)
+   set(multiValueArgs SOURCES)
+   cmake_parse_arguments(ENCLAVE
+     "${options}"
+     "${oneValueArgs}"
+     "${multiValueArgs}"
+     ${ARGN})
+
+  # Set up the linker flags exactly as we need them such that the resulting
+  # binary be compatible with OP-TEE's loader.
+  set(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS)
+  set(CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS)
+  set(CMAKE_EXE_EXPORTS_C_FLAG)
+
+  string(REPLACE "gcc" "ld" LINKER ${CMAKE_C_COMPILER})
+  set(CMAKE_C_LINK_EXECUTABLE "${LINKER} <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> <LINK_LIBRARIES> -o <TARGET>")
+  set(CMAKE_CXX_LINK_EXECUTABLE "${LINKER} <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> <LINK_LIBRARIES> -o <TARGET>")
+
+  # Generate linker script from template.
+  string(REPLACE "gcc" "cpp" C_PREPROCESSOR ${CMAKE_C_COMPILER})
+  set(TA_LINKER_SCRIPT ${CMAKE_CURRENT_BINARY_DIR}/ta.ld)
+  set(TA_LINKER_SCRIPT ${TA_LINKER_SCRIPT} PARENT_SCOPE)
+  add_custom_target(${ENCLAVE_TARGET}.ld
+    COMMAND
+      ${C_PREPROCESSOR} -Wp,-P -DASM=1 -DARM64 -nostdinc ${OE_TZ_TA_DEV_KIT_LINKER_SCRIPT_TEMPLATE} > ${TA_LINKER_SCRIPT}
+    SOURCES ${OE_TZ_TA_DEV_KIT_LINKER_SCRIPT_TEMPLATE}
+    BYPRODUCTS ${TA_LINKER_SCRIPT})
+
+  # Ask GCC where is libgcc.
+  execute_process(
+    COMMAND ${CMAKE_C_COMPILER}
+      ${OE_TZ_TA_C_FLAGS}
+      -print-libgcc-file-name
+    OUTPUT_VARIABLE LIBGCC_PATH
+    OUTPUT_STRIP_TRAILING_WHITESPACE)
+  get_filename_component(LIBGCC_PATH ${LIBGCC_PATH} DIRECTORY)
+
+  # Set up the target.
+  add_executable(${ENCLAVE_TARGET} ${ENCLAVE_SOURCES})
+  set_property(TARGET ${ENCLAVE_TARGET} PROPERTY C_STANDARD 99)
+  set_target_properties(${ENCLAVE_TARGET} PROPERTIES OUTPUT_NAME ${ENCLAVE_UUID})
+  set_target_properties(${ENCLAVE_TARGET} PROPERTIES SUFFIX ".elf")
+  add_dependencies(${ENCLAVE_TARGET} ${ENCLAVE_TARGET}.ld)
+  target_include_directories(${ENCLAVE_TARGET} BEFORE PRIVATE
+    ${ENCLAVE_INCLUDES}
+    ${CMAKE_CURRENT_BINARY_DIR}
+    ${CMAKE_CURRENT_SOURCE_DIR}/optee)
+  target_link_libraries(${ENCLAVE_TARGET} oeenclave)
+  if(ENCLAVE_CXX)
+    target_link_libraries(${ENCLAVE_TARGET} oelibcxx)
+  endif()
+
+  # Strip unneeded bits.
+  string(REPLACE "gcc" "objcopy" OBJCOPY ${CMAKE_C_COMPILER})
+  add_custom_target(${ENCLAVE_TARGET}.stripped.elf
+    COMMAND
+      ${OBJCOPY}
+        --strip-unneeded $<TARGET_FILE:${ENCLAVE_TARGET}>
+        $<TARGET_FILE_DIR:${ENCLAVE_TARGET}>/${ENCLAVE_UUID}.stripped.elf
+    BYPRODUCTS $<TARGET_FILE_DIR:${ENCLAVE_TARGET}>/${ENCLAVE_UUID}.stripped.elf)
+  add_dependencies(${ENCLAVE_TARGET}.stripped.elf ${ENCLAVE_TARGET})
+
+  # Sign the TA with the given key, or with the default key if none was given.
+  if(NOT ENCLAVE_KEY)
+    set(ENCLAVE_KEY ${OE_TZ_TA_DEV_KIT_DEFAULT_SIGNING_KEY})
+  endif()
+  add_custom_target(${ENCLAVE_TARGET}.ta ALL
```

Comment:
Should you have a dependency on `OE_TZ_TA_DEV_KIT_SIGN_TOOL` or does that variable resolve to an executable target in the CMake dependency graph?

Reply:
`OE_TZ_TA_DEV_KIT_SIGN_TOOL` is a Python script in the OP-TEE OS submodule; it's always there.